File: tool/fetch_qul_data.py
import sys
import os
import json
import urllib.request

def main():
    if len(sys.argv) < 2:
        print("Usage: python tool/fetch_qul_data.py <chapter_number> [reciter_id]")
        sys.exit(1)
        
    chapter_number = sys.argv[1]
    reciter_id = sys.argv[2] if len(sys.argv) > 2 else "7"
    
    output_dir = "assets/mock_recitations"
    os.makedirs(output_dir, exist_ok=True)
    
    print(f"Fetching data for Chapter {chapter_number}, Reciter {reciter_id}...")
    
    url = f"https://api.quran.com/api/v4/chapter_recitations/{reciter_id}/{chapter_number}?segments=true"
    
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    
    try:
        with urllib.request.urlopen(req) as response:
            data = json.loads(response.read().decode())
    except Exception as e:
        print(f"Failed to fetch API data: {e}")
        sys.exit(1)
        
    audio_file = data.get("audio_file", {})
    audio_url = audio_file.get("audio_url")
    
    if not audio_url:
        print("No audio_url found in response.")
        sys.exit(1)
        
    if not audio_url.startswith("http"):
        # Could be missing domain
        if audio_url.startswith("//"):
            full_audio_url = "https:" + audio_url
        else:
            full_audio_url = "https://audio.qurancdn.com/" + audio_url
    else:
        full_audio_url = audio_url
        
    print(f"Downloading audio from {full_audio_url}...")
    
    mp3_path = os.path.join(output_dir, f"{chapter_number}.mp3")
    json_path = os.path.join(output_dir, f"{chapter_number}.json")
    
    try:
        req_audio = urllib.request.Request(full_audio_url, headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req_audio) as response:
            with open(mp3_path, "wb") as out_file:
                out_file.write(response.read())
    except Exception as e:
        print(f"Failed to download audio: {e}")
        sys.exit(1)
        
    with open(json_path, "w", encoding="utf-8") as f:
        json.dump(audio_file, f, ensure_ascii=False, indent=2)
        
    print(f"Successfully saved MP3 and JSON data to {output_dir}")

if __name__ == "__main__":
    main()

File: tool/fetch_qul_data.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;

Future<void> main(List<String> arguments) async {
  if (arguments.isEmpty) {
    print('Usage: dart run tool/fetch_qul_data.dart <chapter_number> [reciter_id]');
    print('Example: dart run tool/fetch_qul_data.dart 1 7');
    exit(1);
  }

  final chapterNumber = int.parse(arguments[0]);
  final reciterId = arguments.length > 1 ? int.parse(arguments[1]) : 7; // 7 is Mishary Al-Afasy by default

  final outputDir = Directory('assets/mock_recitations');
  if (!await outputDir.exists()) {
    await outputDir.create(recursive: true);
  }

  print('Fetching data for Chapter $chapterNumber, Reciter $reciterId...');

  // Using Quran.com v4 API which serves QUL segment data
  final uri = Uri.parse('https://api.quran.com/api/v4/chapter_recitations/$reciterId/$chapterNumber?segments=true');
  final response = await http.get(uri);

  if (response.statusCode != 200) {
    print('Failed to fetch API data: ${response.statusCode}');
    print(response.body);
    exit(1);
  }

  final data = jsonDecode(response.body);
  final audioUrl = data['audio_file']['audio_url'];
  
  // Clean up URL if it lacks scheme
  final fullAudioUrl = audioUrl.startsWith('http') ? audioUrl : 'https://api.quran.com$audioUrl'; // Sometimes it needs this or audio.qurancdn.com

  print('Downloading audio from $fullAudioUrl...');
  final audioResponse = await http.get(Uri.parse(audioUrl.startsWith('http') ? audioUrl : 'https://$audioUrl'));
  
  if (audioResponse.statusCode != 200) {
    print('Failed to download audio: ${audioResponse.statusCode}');
    // Attempt alternative CDN logic if needed
    final altResponse = await http.get(Uri.parse('https://audio.qurancdn.com/$audioUrl'));
    if (altResponse.statusCode == 200) {
       final audioFile = File('${outputDir.path}/$chapterNumber.mp3');
       await audioFile.writeAsBytes(altResponse.bodyBytes);
    } else {
       print('Failed alternative CDN too.');
       exit(1);
    }
  } else {
    final audioFile = File('${outputDir.path}/$chapterNumber.mp3');
    await audioFile.writeAsBytes(audioResponse.bodyBytes);
  }

  // Save the JSON segments
  final jsonFile = File('${outputDir.path}/$chapterNumber.json');
  await jsonFile.writeAsString(jsonEncode(data['audio_file']));

  print('Successfully saved MP3 and JSON data to ${outputDir.path}');
}
